[MediaStream] Add support for whiteBalanceMode

eric-carlson opened the pull request:
#### 351ec2749d2563ff7aa631bcdcabcc58240a3889
<pre>
[MediaStream] Add support for whiteBalanceMode
<a href="https://bugs.webkit.org/show_bug.cgi?id=261602">https://bugs.webkit.org/show_bug.cgi?id=261602</a>
rdar://115552800

Reviewed by Youenn Fablet.

Add support for MediaTrackCapabilities `whiteBalanceMode` and expose &apos;whiteBalanceMode&apos; as
a supported constraint.

Update mock camera to support whiteBalanceMode and add new test.

* LayoutTests/fast/mediastream/MediaStreamTrack-getCapabilities-expected.txt:
* LayoutTests/fast/mediastream/MediaStreamTrack-getCapabilities.html:
* LayoutTests/fast/mediastream/mediastreamtrack-video-white-balance-mode-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-white-balance-mode.html: Added.
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
(WebCore::MediaDevices::getSupportedConstraints):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
(WebCore::capabilityStringVector):
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/Modules/mediastream/MeteringMode.h: New.
* Source/WebCore/Modules/mediastream/MeteringMode.idl: New.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::whiteBalanceMode const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::supportsConstraints):
(WebCore::RealtimeMediaSource::setWhiteBalanceMode):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceCapabilities::whiteBalanceModes const):
(WebCore::RealtimeMediaSourceCapabilities::setWhiteBalanceModes):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::RealtimeMediaSourceCenter):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::convertEnumerationToString):
(WebCore::RealtimeMediaSourceSettings::facingMode): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::whiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::setWhiteBalanceMode):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsWhiteBalanceMode):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::whiteBalanceModeFromMeteringMode):
(WebCore::meteringModeFromAVCaptureWhiteBalanceMode):
(WebCore::AVVideoCaptureSource::settingsDidChange):
(WebCore::supportedWhiteBalanceModes):
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::capabilities):
(WebCore::AVVideoCaptureSource::updateWhiteBalanceMode):
* Source/WebCore/platform/mock/MockMediaDevice.h:
(WebCore::MockCameraProperties::encode const):
(WebCore::MockCameraProperties::decode):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/268102@main">https://commits.webkit.org/268102@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acdd31c4ef41e4adf4fdb84e7e04665de16edea9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18695 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19036 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19639 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20559 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17501 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18892 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22345 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19177 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19339 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18920 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19073 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16280 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21437 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16296 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17040 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23492 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17321 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17212 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21386 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17810 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/15120 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/22065 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16869 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/16872 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5383 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4441 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21236 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/23310 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17649 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5238 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
```diff
@@ -27,6 +27,7 @@
 
 #if ENABLE(MEDIA_STREAM)
 
+#include "JSMeteringMode.h"
```
Why do we need JSMeteringMode.h and not JSVideoFacingMode.h

eric-carlson replied:
There is no JSVideoFacingMode.h because we don't define the FacingMode enum in idl. This patch was already big enough when I found that so I decided to do it in a followup.

youennf commented on Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
```diff
@@ -168,11 +153,11 @@ OptionSet<RealtimeMediaSourceSettings::Flag> RealtimeMediaSourceSettings::differ
 String convertEnumerationToString(VideoFacingMode enumerationValue)
```
We can remove this one and use the JSVideoFacingMode.h version?

eric-carlson replied:
I will do that when it exists.

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -257,6 +295,21 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+static Vector<MeteringMode> supportedWhiteBalanceModes(AVCaptureDevice* device)
+{
+    Vector<MeteringMode> whiteBalanceModes;
+    whiteBalanceModes.reserveCapacity(3);
```
reserveInitialCapacity.

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -257,6 +295,21 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+    whiteBalanceModes.reserveCapacity(3);
+
+    if ([device isWhiteBalanceModeSupported:AVCaptureWhiteBalanceModeLocked])
+        whiteBalanceModes.append(MeteringMode::Manual);
```
uncheckedAppend

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -296,6 +349,11 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
         settings.setZoom(zoom());
     }
 
+    if (supportedWhiteBalanceModes(device()).size()) {
```
Would it read better with !isEmpty() instead?

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -328,6 +386,14 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
     }
 #endif // HAVE(AVCAPTUREDEVICE_MINFOCUSLENGTH)
 
+    Vector<MeteringMode> whiteBalanceModes = supportedWhiteBalanceModes(videoDevice);
```
auto

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -328,6 +386,14 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
 #endif // HAVE(AVCAPTUREDEVICE_MINFOCUSLENGTH)
 
+    Vector<MeteringMode> whiteBalanceModes = supportedWhiteBalanceModes(videoDevice);
+    if (whiteBalanceModes.size()) {
```
Ditto for isEmpty?

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -92,6 +92,41 @@ static dispatch_queue_t globaVideoCaptureSerialQueue()
     return globalQueue;
 }
 
+static std::optional<AVCaptureWhiteBalanceMode> whiteBalanceModeFromMeteringMode(MeteringMode mode)
```
Why optional, can we just return a default value, say AVCaptureWhiteBalanceModeAutoWhiteBalance

eric-carlson replied:
True, changed

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -92,6 +92,41 @@ static dispatch_queue_t globaVideoCaptureSerialQueue()
+        return { AVCaptureWhiteBalanceModeContinuousAutoWhiteBalance };
+        break;
+    }
+
```
ASSERT?

eric-carlson replied:
Fixed

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -506,6 +572,28 @@ static inline IntDegrees sensorOrientation(AVCaptureVideoOrientation videoOrient
+    auto* device = this->device();
+    @try {
+        if ([device lockForConfiguration:&error])
+            device.whiteBalanceMode = mode.value();
```
We are not doing so right now, but should we log the error if we are not able to lock?

eric-carlson replied:
Good idea, fixed.

youennf reviewed (approved):
LGTM.
We might need to implement a getUserMedia specific step which disallows using exact for extension constraints:
```
If CS contains a member that is a [required constraint](https://w3c.github.io/mediacapture-main/#dfn-required-constraints) and whose name is not in the list of [allowed required constraints for device selection](https://w3c.github.io/mediacapture-main/#dfn-allowed-required-constraints-for-device-selection), then [reject](https://webidl.spec.whatwg.org/#reject) p with a [TypeError](https://webidl.spec.whatwg.org/#exceptiondef-typeerror), and abort these steps.
```

eric-carlson committed:
[MediaStream] Add support for whiteBalanceMode
https://bugs.webkit.org/show_bug.cgi?id=261602
rdar://115552800

Reviewed by Youenn Fablet.

Add support for MediaTrackCapabilities `whiteBalanceMode` and expose 'whiteBalanceMode' as
a supported constraint.

Update mock camera to support whiteBalanceMode and add new test.

* LayoutTests/fast/mediastream/MediaStreamTrack-getCapabilities-expected.txt:
* LayoutTests/fast/mediastream/MediaStreamTrack-getCapabilities.html:
* LayoutTests/fast/mediastream/mediastreamtrack-video-white-balance-mode-expected.txt: Added.
* LayoutTests/fast/mediastream/mediastreamtrack-video-white-balance-mode.html: Added.
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/mediastream/MediaDevices.cpp:
(WebCore::hasInvalidGetDisplayMediaConstraint):
(WebCore::MediaDevices::getSupportedConstraints):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::getSettings const):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.idl:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.cpp:
(WebCore::capabilityStringVector):
(WebCore::toMediaTrackCapabilities):
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.h:
* Source/WebCore/Modules/mediastream/MediaTrackCapabilities.idl:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.cpp:
(WebCore::convertToInternalForm):
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackConstraints.idl:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.h:
* Source/WebCore/Modules/mediastream/MediaTrackSupportedConstraints.idl:
* Source/WebCore/Modules/mediastream/MeteringMode.h: New.
* Source/WebCore/Modules/mediastream/MeteringMode.idl: New.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mediastream/MediaConstraints.cpp:
(WebCore::MediaTrackConstraintSetMap::filter const):
(WebCore::MediaTrackConstraintSetMap::set):
* Source/WebCore/platform/mediastream/MediaConstraints.h:
(WebCore::MediaTrackConstraintSetMap::whiteBalanceMode const):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::fitnessDistance):
(WebCore::RealtimeMediaSource::applyConstraint):
(WebCore::RealtimeMediaSource::supportsConstraint):
(WebCore::RealtimeMediaSource::supportsConstraints):
(WebCore::RealtimeMediaSource::setWhiteBalanceMode):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCapabilities.h:
(WebCore::RealtimeMediaSourceCapabilities::RealtimeMediaSourceCapabilities):
(WebCore::RealtimeMediaSourceCapabilities::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceCapabilities::whiteBalanceModes const):
(WebCore::RealtimeMediaSourceCapabilities::setWhiteBalanceModes):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceCenter.cpp:
(WebCore::RealtimeMediaSourceCenter::RealtimeMediaSourceCenter):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.cpp:
(WebCore::RealtimeMediaSourceSettings::convertFlagsToString):
(WebCore::convertEnumerationToString):
(WebCore::RealtimeMediaSourceSettings::facingMode): Deleted.
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSettings.h:
(WebCore::RealtimeMediaSourceSettings::allFlags):
(WebCore::RealtimeMediaSourceSettings::RealtimeMediaSourceSettings):
(WebCore::RealtimeMediaSourceSettings::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::whiteBalanceMode const):
(WebCore::RealtimeMediaSourceSettings::setWhiteBalanceMode):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.cpp:
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsConstraint const):
* Source/WebCore/platform/mediastream/RealtimeMediaSourceSupportedConstraints.h:
(WebCore::RealtimeMediaSourceSupportedConstraints::RealtimeMediaSourceSupportedConstraints):
(WebCore::RealtimeMediaSourceSupportedConstraints::supportsWhiteBalanceMode const):
(WebCore::RealtimeMediaSourceSupportedConstraints::setSupportsWhiteBalanceMode):
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::whiteBalanceModeFromMeteringMode):
(WebCore::meteringModeFromAVCaptureWhiteBalanceMode):
(WebCore::AVVideoCaptureSource::settingsDidChange):
(WebCore::supportedWhiteBalanceModes):
(WebCore::AVVideoCaptureSource::settings):
(WebCore::AVVideoCaptureSource::capabilities):
(WebCore::AVVideoCaptureSource::updateWhiteBalanceMode):
* Source/WebCore/platform/mock/MockMediaDevice.h:
(WebCore::MockCameraProperties::encode const):
(WebCore::MockCameraProperties::decode):
* Source/WebCore/platform/mock/MockRealtimeMediaSourceCenter.cpp:
(WebCore::defaultDevices):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::MockRealtimeVideoSource::capabilities):
(WebCore::MockRealtimeVideoSource::settings):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/268102@main
